### PR TITLE
Add tutorial for saving and loading `tf.Model`

### PR DESCRIPTION
**Pull request by caisq**
Towards: https://github.com/tensorflow/tfjs/issues/13

This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-website/147)

### CODE REVIEW

File: source/tutorials/model-save-load.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+  the path is just an arbitrary string that
+  uniquely identifies the model being saved. It will be used, for example,
+  when you load model back from local storage.
+- The `save` method is asynchronous, so you need to use `then` or `await` if
+  its completion forms the precondition of other actions.
+
+The table below lists all currently supported destinations of saving models an
+their respecitve schemes and examples.
+
+| Savig Destination          | Scheme string     | Code example                                     |
```
Comment (jgautsch) [inline source/tutorials/model-save-load.md]: typo- "Savig"

File: source/tutorials/model-save-load.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+filename prefix to be downloaded from the browser (note: some browsers
+require users to grant permissions before more than one file can be downloaded
+at the same time.):
+
+  1. A text JSON file named `my-model-1.json` carrying topology (architecture)
+     of the model in its `modelTopology` field and a manifest of weights in its
+     `weightsManifest` field.
+  2. A binary file carrying the weight values named `my-model-1.weights.bin`.
+
+These two files are in the same format as the artifacts converted from Keas
```
Comment (jgautsch) [inline source/tutorials/model-save-load.md]: typo- "Keas" -> "Keras"

File: source/tutorials/model-save-load.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    {method: 'PUT',
+     headers: {'header_key_1': 'header_value_1'}}));
+```
+
+## Loading `tf.Model`s.
+
+The ability to save `tf.Model`s will not be useful if the models can't be
+loaded back afterwards. Loading models are done with `tf.loadModel`, with a
+scheme-based URL-like arguments that symmetrical to `tf.Model.save` in most
+cases. The table below gives a summary of the supported loading routes:
```
Comment (jgautsch) [inline source/tutorials/model-save-load.md]: > Loading models are done with `tf.loadModel`, with a scheme-based URL-like arguments that symmetrical to `tf.Model.save` in most cases.

grammar

File: source/tutorials/model-save-load.md (hunk 1 above)
Comment (caisq) [reply source/tutorials/model-save-load.md]: Done.

File: source/tutorials/model-save-load.md (hunk 2 above)
Comment (caisq) [reply source/tutorials/model-save-load.md]: Done.

File: source/tutorials/model-save-load.md (hunk 3 above)
Comment (caisq) [reply source/tutorials/model-save-load.md]: Done.

**Comment by nsthorat** [comment]
:lgtm_strong:

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[source/tutorials/index.md, line 34 at r2](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCVFuCpCdqlQ7YaqxXL:-LCVFuCqBWKi6XF8MM5z:bykyglh) ([raw file](https://github.com/tensorflow/tfjs-website/blob/9f74bf31c05ec864fed7a29392e8233d557f0ffb/source/tutorials/index.md#L34)):*
> ```Markdown
> <h2><a href="./model-save-load.html">Saving and Loading `tf.Model`s</a></h2>
> 
> This tutorial explains how to save `tf.Model`s to various destinations such as browser IndexedDB and load them back.
> ```

say local storage here since IndexedDB is a little newer

---

*[source/tutorials/model-save-load.md, line 6 at r2](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCVH2UN9pZHQJx1KTYd:-LCVH2UN9pZHQJx1KTYe:bfkusbv) ([raw file](https://github.com/tensorflow/tfjs-website/blob/9f74bf31c05ec864fed7a29392e8233d557f0ffb/source/tutorials/model-save-load.md#L6)):*
> ```Markdown
> ---
> 
> # Saving and Loading `tf.Model`s
> ```

This renders kind of funny, let's remove the markdown and just say "Saving and loading layers models"

---

*[source/tutorials/model-save-load.md, line 8 at r2](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCVGgY92lHQOFGR_W-q:-LCVGgY92lHQOFGR_W-r:b-1zak9t) ([raw file](https://github.com/tensorflow/tfjs-website/blob/9f74bf31c05ec864fed7a29392e8233d557f0ffb/source/tutorials/model-save-load.md#L8)):*
> ```Markdown
> # Saving and Loading `tf.Model`s
> 
> No deep learning API is complete without the capability to save and load
> ```

I think this is a little too much fluff, what about just jumping in and saying "This tutorial will show you how to load and save your models" or something along those lines

---

*[source/tutorials/model-save-load.md, line 22 at r2](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCVGp9fASJvN5ojTpoE:-LCVGp9fASJvN5ojTpoF:b-13kqll) ([raw file](https://github.com/tensorflow/tfjs-website/blob/9f74bf31c05ec864fed7a29392e8233d557f0ffb/source/tutorials/model-save-load.md#L22)):*
> ```Markdown
> 
> Saving and loading `tf.Model`s are done by the `tf.Model.save` and
> `tf.loadModel` methods, respectively. We designed these APIs to be similar to
> ```

I would avoid saying this is similar to keras, this is not really important (and very high up in the tutorial!)

---

*[source/tutorials/model-save-load.md, line 25 at r2](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCVGzQ6Bv3paZwlbtPI:-LCVGzQ6Bv3paZwlbtPJ:bmgm4h9) ([raw file](https://github.com/tensorflow/tfjs-website/blob/9f74bf31c05ec864fed7a29392e8233d557f0ffb/source/tutorials/model-save-load.md#L25)):*
> ```Markdown
> [the save and load_model API](https://keras.io/getting-started/faq/#how-can-i-save-a-keras-model)
> of Keras. But the browser environment is quite different from the backend environment
> on which staple deep learning frameworks like Keras run, particularly in the
> ```

 reduce usage of keras here to reduce confusion :)

---

*[source/tutorials/model-save-load.md, line 37 at r2](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCZHY6p7HacfoM-9vOz:-LCZHY6p7HacfoM-9vP-:bc8yzcs) ([raw file](https://github.com/tensorflow/tfjs-website/blob/9f74bf31c05ec864fed7a29392e8233d557f0ffb/source/tutorials/model-save-load.md#L37)):*
> ```Markdown
> 
> Let's begin with the most basic, hassle-free way of saving a `tf.Model`: to
> the local storage of the web browser. Local storage is a standard
> ```

capitalize Local Storage here and everywhere

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-:-LCZKNWSBEVjZ1zOJ4T0:b-jbsf8z)*

**Comment by augbog** [comment]
Review status: 0 of 2 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[source/tutorials/model-save-load.md, line 32 at r2](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCWJ0cyDFJOYMUoEeFh:-LCWJ0cyDFJOYMUoEeFi:by82xdg) ([raw file](https://github.com/tensorflow/tfjs-website/blob/9f74bf31c05ec864fed7a29392e8233d557f0ffb/source/tutorials/model-save-load.md#L32)):*
> ```Markdown
> > models in the tfjs-layers API). Saving and loading `tf.FrozenModel`s (i.e.,
> > models loaded from TensoFlow SavedModels) are not supported yet and is being
> > activately worked on.
> ```

nit: actively* 

Also it might be good to link to a formal Github issue if there is one tracking it :)

---

*[source/tutorials/model-save-load.md, line 107 at r2](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCZhRv99q-gys9_Se3J:-LCWKY39D4JTbJ4gN0VT:b-lddidy) ([raw file](https://github.com/tensorflow/tfjs-website/blob/9f74bf31c05ec864fed7a29392e8233d557f0ffb/source/tutorials/model-save-load.md#L107)):*
> ```Markdown
> If `tf.Model.save` is called with an HTTP/HTTPS URL, the topology and weights of the
> model will be sent to the specified HTTP server via a
> [POST](https://en.wikipedia.org/wiki/POST_(HTTP)) request. The body of the POST
> ```

nit: consider opting for MDN documentation on [POST](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST)? Not required though :) awesome section btw

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-website/147)*

**Comment by caisq** [comment]
Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[source/tutorials/index.md, line 34 at r2](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCVFuCpCdqlQ7YaqxXL:-LC_axO--NvEf35nZ_ss:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-website/blob/9f74bf31c05ec864fed7a29392e8233d557f0ffb/source/tutorials/index.md#L34)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

say local storage here since IndexedDB is a little newer
</blockquote></details>

Done.

---

*[source/tutorials/model-save-load.md, line 6 at r2](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCVH2UN9pZHQJx1KTYd:-LC_cn1k9RWV1sdCPmeZ:bakbr65) ([raw file](https://github.com/tensorflow/tfjs-website/blob/9f74bf31c05ec864fed7a29392e8233d557f0ffb/source/tutorials/model-save-load.md#L6)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

This renders kind of funny, let's remove the markdown and just say "Saving and loading layers models"
</blockquote></details>

I changed it to something like

Saving tf.Model
Loading tf.Model

---

*[source/tutorials/model-save-load.md, line 8 at r2](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCVGgY92lHQOFGR_W-q:-LC_f-88AujsjUvrPV5p:b-2h5e9p) ([raw file](https://github.com/tensorflow/tfjs-website/blob/9f74bf31c05ec864fed7a29392e8233d557f0ffb/source/tutorials/model-save-load.md#L8)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

I think this is a little too much fluff, what about just jumping in and saying "This tutorial will show you how to load and save your models" or something along those lines
</blockquote></details>

I made it less fluffy.

---

*[source/tutorials/model-save-load.md, line 22 at r2](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCVGp9fASJvN5ojTpoE:-LC_bEYF0AcukNtIRCV_:b-odyuzv) ([raw file](https://github.com/tensorflow/tfjs-website/blob/9f74bf31c05ec864fed7a29392e8233d557f0ffb/source/tutorials/model-save-load.md#L22)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

I would avoid saying this is similar to keras, this is not really important (and very high up in the tutorial!)
</blockquote></details>

Done. I moved it to the existing tutorial page " TensorFlow.js Layers API for Keras Users"

---

*[source/tutorials/model-save-load.md, line 25 at r2](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCVGzQ6Bv3paZwlbtPI:-LC_bneE2h2FpHp_XOuV:b-a4fvys) ([raw file](https://github.com/tensorflow/tfjs-website/blob/9f74bf31c05ec864fed7a29392e8233d557f0ffb/source/tutorials/model-save-load.md#L25)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

 reduce usage of keras here to reduce confusion :)
</blockquote></details>

Again, moved to "TensorFlow.js Layers API for Keras Users".

---

*[source/tutorials/model-save-load.md, line 32 at r2](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCWJ0cyDFJOYMUoEeFh:-LC_bsUi5LY-PAyA8mXz:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-website/blob/9f74bf31c05ec864fed7a29392e8233d557f0ffb/source/tutorials/model-save-load.md#L32)):*
<details><summary><i>Previously, augbog (Augustus Yuan) wrote…</i></summary><blockquote>

nit: actively* 

Also it might be good to link to a formal Github issue if there is one tracking it :)
</blockquote></details>

Done.

---

*[source/tutorials/model-save-load.md, line 37 at r2](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCZHY6p7HacfoM-9vOz:-LC_c9rtABawAchFpJAq:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-website/blob/9f74bf31c05ec864fed7a29392e8233d557f0ffb/source/tutorials/model-save-load.md#L37)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

capitalize Local Storage here and everywhere
</blockquote></details>

Done.

---

*[source/tutorials/model-save-load.md, line 107 at r2](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCZhRv99q-gys9_Se3J:-LC_cFxz1bN60hpwdGBq:b-l5r1cy) ([raw file](https://github.com/tensorflow/tfjs-website/blob/9f74bf31c05ec864fed7a29392e8233d557f0ffb/source/tutorials/model-save-load.md#L107)):*
<details><summary><i>Previously, augbog (Augustus Yuan) wrote…</i></summary><blockquote>

nit: consider opting for MDN documentation on [POST](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST)? Not required though :) awesome section btw
</blockquote></details>

Done.  Thanks.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-website/147)*

**Comment by bileschi** [comment]
*[source/tutorials/model-save-load.md, line 35 at r3](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCdeUoI2Rcz-Jcg6K2Q:-LCdeUoI2Rcz-Jcg6K2R:bw27vok) ([raw file](https://github.com/tensorflow/tfjs-website/blob/71a2fb717db7c4ce79a9094e4e8695de029bb1d2/source/tutorials/model-save-load.md#L35)):*
> ```
> Suppose you have a `tf.Model` object called `model`. Be it created by using the
> Layers API from scratch or loaded/fine-tuned from a pretrained Keras model,
> you can save it to Local Storage with one line of code:
> ```

I think this would be clearer by striking the aside, i.e.:

> ```
> If you have a `tf.Model` object called `model`, you can save it to Local Storage with one line of code:
> ```

If you strongly wish to highlight that models can be built from scratch or loaded, maybe add it to the things worth pointing out below?

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-website/147)*

**Comment by bileschi** [comment]
:lgtm_strong:

---

Review status: 0 of 3 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[source/tutorials/model-save-load.md, line 42 at r3](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCdfHTu3-gq4ApHqWW5:-LCdfHTu3-gq4ApHqWW6:bxp7tgi) ([raw file](https://github.com/tensorflow/tfjs-website/blob/71a2fb717db7c4ce79a9094e4e8695de029bb1d2/source/tutorials/model-save-load.md#L42)):*
> ```
> A few things are worth pointing out:
> ```
Are there any kinds of models which can *not* be saved in this way?  If so, please highlight it here.  If not, please highlight that *all* kinds of models can be saved in this way.

---

*[source/tutorials/model-save-load.md, line 81 at r3](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCdfhQqCr56Em1hWPW2:-LCdfhQqCr56Em1hWPW3:b-yft8rv) ([raw file](https://github.com/tensorflow/tfjs-website/blob/71a2fb717db7c4ce79a9094e4e8695de029bb1d2/source/tutorials/model-save-load.md#L81)):*
> ```
> will cause two files
> ```
This is the first time that "two files" is mentioned, and users will be surprised that they are downloading one model and getting two files.   Perhaps ease the reader into it with:

```
will download the model topology as `my-model-1.json` and the weights as `my-model-1.weights.bin`.    (note: some browsers ....)

In more detail:
 1.  The text JSON file named  `my-model-1.json` contains ...
```


---

*[source/tutorials/model-save-load.md, line 149 at r3](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCdhYuF125npzVWWU5l:-LCdhYuF125npzVWWU5m:b-cxacuv) ([raw file](https://github.com/tensorflow/tfjs-website/blob/71a2fb717db7c4ce79a9094e4e8695de029bb1d2/source/tutorials/model-save-load.md#L149)):*
> ```
> hence it merit no further discussion other
> than the fact that a model must have been saved previously to the path
> following the scheme.
> ```
Maybe remove this phrase.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-:-LCdiTbr2InMODXjnBX1:b-jbsf8z)*

**Comment by bileschi** [comment]
*[source/tutorials/model-save-load.md, line 210 at r3](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCdiVrp470X4o8dWGQi:-LCdiVrp470X4o8dWGQj:b-i3m2gz) ([raw file](https://github.com/tensorflow/tfjs-website/blob/71a2fb717db7c4ce79a9094e4e8695de029bb1d2/source/tutorials/model-save-load.md#L210)):*
<details><summary><i alt="summary">Quoted 10 lines of code…</i></summary>

> ```
> // Copy a model in Local Storage, from an existing path to a new path.
> tf.io.browserLocalStorageManager().copyModel('my-model', 'cloned-model');
>  
> // Copy a model in IndexedDB, from an existing path to a new path.
> tf.io.browserIndexedDBManager().copyModel('my-model', 'cloned-model');
>  
> // Delete a model from Local Storage.
> tf.io.browserLocalStorageManager().deleteModel('clone-model');
>  
> // Delete a model from IndexedDB.
> ```
</details>

Dont forget to update this once the manager API is updated.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-website/147)*

**Comment by caisq** [comment]
Review status: 0 of 3 files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[source/tutorials/model-save-load.md, line 42 at r3](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCdfHTu3-gq4ApHqWW5:-LCeLsxj-Pal7Z2BeeRO:bbt5hp) ([raw file](https://github.com/tensorflow/tfjs-website/blob/71a2fb717db7c4ce79a9094e4e8695de029bb1d2/source/tutorials/model-save-load.md#L42)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> A few things are worth pointing out:
> ```
Are there any kinds of models which can *not* be saved in this way?  If so, please highlight it here.  If not, please highlight that *all* kinds of models can be saved in this way.

</blockquote></details>

All tfjs-layers models can be saved this way. Done.

---

*[source/tutorials/model-save-load.md, line 81 at r3](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCdfhQqCr56Em1hWPW2:-LCeLj4v-SvCGDBIcwP4:b3zjwki) ([raw file](https://github.com/tensorflow/tfjs-website/blob/71a2fb717db7c4ce79a9094e4e8695de029bb1d2/source/tutorials/model-save-load.md#L81)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> will cause two files
> ```
This is the first time that "two files" is mentioned, and users will be surprised that they are downloading one model and getting two files.   Perhaps ease the reader into it with:

```
will download the model topology as `my-model-1.json` and the weights as `my-model-1.weights.bin`.    (note: some browsers ....)

In more detail:
 1.  The text JSON file named  `my-model-1.json` contains ...

</blockquote></details>

Done. I agree with your concern. I reworded this paragraph in a different way. 

---

*[source/tutorials/model-save-load.md, line 149 at r3](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCdhYuF125npzVWWU5l:-LCeLQtQ89heGNPM_LvU:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-website/blob/71a2fb717db7c4ce79a9094e4e8695de029bb1d2/source/tutorials/model-save-load.md#L149)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> hence it merit no further discussion other
> than the fact that a model must have been saved previously to the path
> following the scheme.
> ```
Maybe remove this phrase.
</blockquote></details>

Done.

---

*[source/tutorials/model-save-load.md, line 210 at r3](https://reviewable.io/reviews/tensorflow/tfjs-website/147#-LCdiVrp470X4o8dWGQi:-LCe9ICdAOmOt9fDfqP_:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-website/blob/71a2fb717db7c4ce79a9094e4e8695de029bb1d2/source/tutorials/model-save-load.md#L210)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> // Copy a model in Local Storage, from an existing path to a new path.
> tf.io.browserLocalStorageManager().copyModel('my-model', 'cloned-model');
>  
> // Copy a model in IndexedDB, from an existing path to a new path.
> tf.io.browserIndexedDBManager().copyModel('my-model', 'cloned-model');
>  
> // Delete a model from Local Storage.
> tf.io.browserLocalStorageManager().deleteModel('clone-model');
>  
> // Delete a model from IndexedDB.
> ```
Dont forget to update this once the manager API is updated.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-website/147)*
<!-- Sent from Reviewable.io -->

**Comment by bileschi** [comment]
Reviewed 1 of 1 files at r4.
Review status: 1 of 3 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-website/147)*

**Pull request merged**